test/integration/base_integration_test.cc

```diff
@@ -694,4 +698,66 @@ AssertionResult BaseIntegrationTest::compareDeltaDiscoveryRequest(
   return AssertionSuccess();
 }
 
+// Attempt to heuristically discover missing tag-extraction rules when new stats are added.
+// This is done by looking through the entire config for fields named `stat_prefix`, and then
+// validating that those values do not appear in the tag-extracted name of any stat.
+//
+// To add a rule, see `source/common/config/well_known_names.cc`.
+//
+// This is done in all integration tests because it is testing new stats and scopes that are created
+// for which the author isn't aware that tag extraction rules need to be written, and thus the
+// author wouldn't think to write tests for that themselves.
+void BaseIntegrationTest::checkForMissingTagExtractionRules() {
+  BufferingStreamDecoderPtr response = IntegrationUtil::makeSingleRequest(
+      test_server_->adminAddress(), "GET", "/config_dump", "", Http::CodecType::HTTP1);
+  EXPECT_TRUE(response->complete());
+  if (!response->complete()) {
+    // Allow the rest of the test to complete for better diagnostic information about the failure.
+    return;
+  }
+
+  EXPECT_EQ("200", response->headers().getStatusValue());
+  Json::ObjectSharedPtr json;
+  try {
+    json = Json::Factory::loadFromString(response->body());
+  } catch (const std::exception&) {
+    // There are a few integration tests that result in the output being an error message instead of
+    // valid JSON. Skip this test for those cases.
+    return;
+  }
+  std::vector<std::string> stat_prefixes;
+  Json::ObjectCallback find_stat_prefix = [&](const std::string& name,
+                                              const Json::Object& root) -> bool {
+    if (name == "stat_prefix") {
+      auto prefix = root.asString();
+      if (!prefix.empty()) {
+        stat_prefixes.push_back(prefix);
+      }
+    } else if (root.isObject()) {
+      root.iterate(find_stat_prefix);
+    } else if (root.isArray()) {
+      std::vector<Json::ObjectSharedPtr> elements = root.asObjectArray();
+      for (const auto& element : elements) {
+        find_stat_prefix("", *element);
+      }
+    }
+    return true;
+  };
+  find_stat_prefix("", *json);
+  ENVOY_LOG_MISC(debug, "discovered stat_prefixes {}", stat_prefixes);
+
+  auto check_metric = [&](auto& metric) {
+    // Validate that the `stat_prefix` string doesn't appear in the tag-extracted name, indicating
+    // that it wasn't extracted.
+    const std::string tag_extracted_name = metric.tagExtractedName();
```

Comment:
you didn't want to iterate over the tags instead? Maybe explain why in a code comment?